lib/services/services_linux.c

```diff
@@ -590,13 +590,27 @@ log_op_output(svc_action_t *op)
     free(prefix);
 }
 
+/*!
+ * \internal
+ * \brief Process the completion of an asynchronous child process
+ *
+ * \param[in] p         Child process that completed
+ * \param[in] pid       Process ID of child
+ * \param[in] core      (unused)
+ * \param[in] signo     Signal that interrupted child, if any
+ * \param[in] exitcode  Exit status of child process
+ */
 static void
-operation_finished(mainloop_child_t * p, pid_t pid, int core, int signo, int exitcode)
+async_action_complete(mainloop_child_t *p, pid_t pid, int core, int signo,
+                      int exitcode)
 {
     svc_action_t *op = mainloop_child_userdata(p);
 
     mainloop_clear_child_userdata(p);
-    CRM_ASSERT(op->pid == pid);
+    CRM_CHECK(op->pid == pid,
+              op->rc = services__generic_error(op);
+              op->status = PCMK_EXEC_ERROR;
+              return);
 
```

Comment:
did that actually happen? otherwise wouldn't we rather call ASSERT if it shouldn't be possible instead of proceeding with an overwritten stack or something that could be the reason for such 'impossible' things.

Reply:
I've never seen it happen. My general attitude is that callers should handle errors, and libraries should just report them, but it's a blurry line. Here, I was thinking that the executor shouldn't crash if one action is somehow messed up, but that's a good point about such a problem possibly indicating larger issues. The next PR will add an exit reason here explaining what happened, so it becomes more useful at that point. I can see either approach though. Thoughts?

Reply:
Guess with an exit reason that indicates such unexpected stuff we can hand it over to the caller or think about it once again.
When it is something where we expect things like stack-overwrites or so we might want it to be caught as near to where it happened if possible - better for hunting down + prevents further danger